### PR TITLE
Add OTel tracing bucket

### PR DESCRIPTION
**Pull request by kakkoyun**
- Refactor client.NewBucket. This changes the behaviour of `client.NewBucket`. Now it returns, uninstrumented and untraced bucket.
- Add OpenTelemetry TracingBucket.

Now you can combine `objstore.WrapWithMetrics` and `tracing/{opentelemetry,opentracing}.WrapWithTraces` to have old behavior.

Signed-off-by: Kemal Akkoyun <kakkoyun@gmail.com>

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

- Add OTel Tracing bucket
- Introduce tracing package
- Refactor client.NewBucket
- Introduce client.NewInstrumentedBucket
- Add examples

### CODE REVIEW

File: client/factory.go [hunk 1]
```diff
@@ -91,5 +91,11 @@ func NewBucket(logger log.Logger, confContentYaml []byte, reg prometheus.Registe
 		return nil, errors.Wrap(err, fmt.Sprintf("create %s client", bucketConf.Type))
 	}
 
-	return objstore.NewTracingBucket(objstore.BucketWithMetrics(bucket.Name(), objstore.NewPrefixedBucket(bucket, bucketConf.Prefix), reg)), nil
+	return objstore.NewPrefixedBucket(bucket, bucketConf.Prefix), nil
+}
+
+// NewInstrumentedBucket initializes and returns new object storage clients.
+// NOTE: confContentYaml can contain secrets.
```
Comment (brancz) [inline client/factory.go:98]: this note is not relevant here

File: README.md [hunk 2]
```diff
@@ -643,18 +644,19 @@ You can also include any of the optional configuration just like the example in
...
 
 To configure HuaweiCloud Account to use OBS as storage store you need to set these parameters in YAML format stored in a file:
 
 ```yaml mdox-exec="go run scripts/cfggen/main.go --name=cos.Config"
-type: OBS                    
-config:                      
-  bucket: ""                 
-  endpoint: ""               
-  access_key: ""             
+type: COS
```
Comment (brancz) [inline README.md:652]: is this correct? It talks about Huawei OBS above and then COS?

**Review by brancz — COMMENTED** [review]
I'm happy with this. Are you going to take care of fixing Thanos to make sure it doesn't loose instrumentation?

File: README.md (hunk 2 above)
Comment (kakkoyun) [reply README.md:652]: Let me double-check. I think running the linter or some other make action caused this.

**Comment by kakkoyun** [comment]
> I'm happy with this. Are you going to take care of fixing Thanos to make sure it doesn't lose instrumentation?

I'll make sure to do that. Maybe we can use opentracing tracing for the instrumented bucket? WDYT?

File: README.md (hunk 2 above)
Comment (kakkoyun) [reply README.md]: Thanks for catching it. I've fixed the command above that generates this doc.

**Comment by brancz** [comment]
I think just replacing the spots where Thanos just called NewBucket, after this merges replace that with `NewTracingBucket(NewInstrumentedBucket(NewBucket(...)))`. That would be sufficient.

**Comment by kakkoyun** [comment]
> I think just replacing the spots where Thanos just called NewBucket, after this merges replace that with `NewTracingBucket(NewInstrumentedBucket(NewBucket(...)))`. That would be sufficient.

https://github.com/thanos-io/thanos/pull/6507

File: tracing/tracing.go [hunk 3]
```diff
@@ -34,36 +28,11 @@ func ContextWithTracer(ctx context.Context, tracer opentracing.Tracer) context.C
 	return context.WithValue(ctx, tracerKey, tracer)
 }
 
-// tracerFromContext extracts opentracing.Tracer from the given context.
-func tracerFromContext(ctx context.Context) opentracing.Tracer {
+// TracerFromContext extracts opentracing.Tracer from the given context.
+func TracerFromContext(ctx context.Context) opentracing.Tracer {
```
Comment (bwplotka) [inline tracing/tracing.go:32]: Do we need this? Can we move it to opentracing, so "common" tracing pkg is vendor free?

File: client/factory.go [hunk 4]
```diff
@@ -91,5 +91,10 @@ func NewBucket(logger log.Logger, confContentYaml []byte, reg prometheus.Registe
 		return nil, errors.Wrap(err, fmt.Sprintf("create %s client", bucketConf.Type))
 	}
 
-	return objstore.NewTracingBucket(objstore.BucketWithMetrics(bucket.Name(), objstore.NewPrefixedBucket(bucket, bucketConf.Prefix), reg)), nil
+	return objstore.NewPrefixedBucket(bucket, bucketConf.Prefix), nil
```
Comment (bwplotka) [inline client/factory.go:94]: Hm, I think it's worth to mention in change log AND in comment of NewBucket, that this NewBucket no longer instruments and adds span middleware. Callers have to add them on their own now.

File: client/factory.go [hunk 5]
```diff
@@ -91,5 +91,10 @@ func NewBucket(logger log.Logger, confContentYaml []byte, reg prometheus.Registe
...
+}
+
+// NewInstrumentedBucket initializes and returns new object storage clients.
```
Comment (bwplotka) [inline client/factory.go:97]: This is not true - it only wraps with metrics. Should we rather remove `NewInstrumentedBucket` and keep `objstore.BucketWithMetrics`?

File: client/factory.go [hunk 6]
```diff
@@ -91,5 +91,10 @@ func NewBucket(logger log.Logger, confContentYaml []byte, reg prometheus.Registe
...
+func NewInstrumentedBucket(reg prometheus.Registerer, bkt objstore.Bucket) objstore.InstrumentedBucket {
+	return objstore.BucketWithMetrics(bkt.Name(), bkt, reg)
```
Comment (bwplotka) [inline client/factory.go:99]: Renamed this:
```suggestion
objstore.InstrumentedBucket(bkt.Name(), bkt, reg)
```

so we have "InstrumentedBucket", "otel.TracingBucket", "opentracing.TracingBucket"? 🤔

File: client/testconf/fake-gcs.conf.yml [hunk 7]
```diff
@@ -1,3 +0,0 @@
-type: FAKE-GCS
```
Comment (bwplotka) [inline client/testconf/fake-gcs.conf.yml:1]: Why changing, sounds like we remve some test cases which right now are uncovered? 🤔

File: tracing/opentelemetry/opentelemetry.go [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+	"github.com/thanos-io/objstore"
+)
+
+// TracingBucket is a wrapper around objstore.Bucket that adds tracing to all operations using OpenTelemetry.
+type TracingBucket struct {
+	tracer trace.Tracer
+	bkt    objstore.Bucket
+}
+
+func NewTracingBucket(tracer trace.Tracer, bkt objstore.Bucket) objstore.InstrumentedBucket {
```
Comment (bwplotka) [inline tracing/opentelemetry/opentelemetry.go:22]: While we are here we could stop calling it "new" 🤔 . It's not new bucket. It's not even new bucket provider. Perhaps  `otel.TracedBucket(...)`or `WrapWithTracing`

**Review by bwplotka — COMMENTED** [review]
Nice! Some comments, otherwise LGTM 👍🏽

File: client/testconf/fake-gcs.conf.yml (hunk 7 above)
Comment (kakkoyun) [reply client/testconf/fake-gcs.conf.yml:1]: These were not utilized. I haven't tracked them down but my they are left overs from the migration.

**Review by brancz — APPROVED** [review]
lgtm

File: client/factory.go [hunk 9]
```diff
@@ -91,5 +91,10 @@ func NewBucket(logger log.Logger, confContentYaml []byte, reg prometheus.Registe
 		return nil, errors.Wrap(err, fmt.Sprintf("create %s client", bucketConf.Type))
 	}
 
-	return objstore.NewTracingBucket(objstore.BucketWithMetrics(bucket.Name(), objstore.NewPrefixedBucket(bucket, bucketConf.Prefix), reg)), nil
+	return objstore.NewPrefixedBucket(bucket, bucketConf.Prefix), nil
+}
+
+// InstrumentedBucket wraps the given bucket with metrics.
+func InstrumentedBucket(bkt objstore.Bucket, reg prometheus.Registerer) objstore.InstrumentedBucket {
```
[the remark this answers is not in the document]
Comment (kakkoyun) [reply client/factory.go:98]: I'm ok with this approach as well. 
If @bwplotka has nothing else to fix, I'll remove it and merge the PR.

**Comment by kakkoyun** [comment]
Ping @bwplotka, is it ok to merge now?

File: client/factory.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply client/factory.go:98]: Bit confused, I suggested removing this in https://github.com/thanos-io/objstore/pull/61#discussion_r1254437185 and sounds like @brancz also suggest removal of this function. Let's address this before merging (either removing or letting us know why removal does not make sense 🙃

File: client/testconf/fake-gcs.conf.yml (hunk 7 above)
Comment (bwplotka) [reply client/testconf/fake-gcs.conf.yml:1]: 👍🏽

File: tracing/opentelemetry/opentelemetry.go [hunk 10]
```diff
@@ -0,0 +1,190 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package opentelemetry
```
Comment (bwplotka) [inline tracing/opentelemetry/opentelemetry.go:4]: Can be simplified to `otel` if we want:

```suggestion
package otel
```

**Review by bwplotka — APPROVED** [review]
Thanks! Still some unresolved comments, otherwise good to go!

File: tracing/opentelemetry/opentelemetry.go (hunk 10 above)
Comment (kakkoyun) [reply tracing/opentelemetry/opentelemetry.go:4]: Let's keep naming consistent with already existing parts (`opentracing` and `opentelemetry`).

**Review by bwplotka — APPROVED** [review]

File: tracing/opentelemetry/opentelemetry.go [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+	defer func() {
+		if err != nil {
+			span.RecordError(err)
+		}
+	}()
+	return t.bkt.Delete(ctx, name)
+}
+
+func (t TracingBucket) Name() string {
+	return "tracing: " + t.bkt.Name()
```
Comment (bboreham) [inline tracing/opentelemetry/opentelemetry.go:120]: Why add this?

**Pull request merged**